src/app/services/models/annotation.service.ts

```diff
@@ -41,34 +45,103 @@ export class AnnotationService {
             in: tagIds,
           },
         } as any,
-      })
+      }),
     );
   }
 
   private async createAudioRecording(
-    audioEvent: AudioEvent
+    audioEvent: AudioEvent,
   ): Promise<AudioRecording> {
     return new AudioRecording(
       await firstValueFrom(
-        this.audioRecordingsApi.show(audioEvent.audioRecordingId)
-      )
+        this.audioRecordingsApi.show(audioEvent.audioRecordingId),
+      ),
     );
   }
 
-  private createAudioLink(audioEvent: AudioEvent): string {
-    const basePath = `/audio_recordings/${audioEvent.audioRecordingId}/media.flac`;
-    let urlParams =
-      `?audio_event_id=${audioEvent.id}` +
-      `&end_offset=${audioEvent.endTimeSeconds}&start_offset=${audioEvent.startTimeSeconds}`;
+  private createAudioLink(
+    audioEvent: AudioEvent,
+    audioRecording: AudioRecording,
+  ): string {
+    // this is here to get around a rounding bug with range requests in the api
+    // see: https://github.com/QutEcoacoustics/baw-server/issues/681
+    // TODO: remove the rounding patch once the api is fixed
+    const roundedStartTime = Math.ceil(audioEvent.startTimeSeconds);
+    // TODO: check that when rounded the start time doesn't excede the end of
+    // the audio file
+    let safeEventStartTime = roundedStartTime;
+
+    // if the rounded end time will excede the duration of the audio recording
+    // we want to use the duration of the audio recording as the events end
+    // we floor the value of the audio recording because rounding up will
+    // overflow the audio recordings time, and we need to have a round number
+    // to prevent api bugs
+    const roundedEndTime = Math.ceil(audioEvent.endTimeSeconds);
+    const recordingEndTime = Math.floor(audioRecording.durationSeconds);
+    let safeEventEndTime = Math.min(roundedEndTime, recordingEndTime);
+
+    // if the audio event is less than .5 seconds, the api will return an error
+    // when trying to split the audio
+    // therefore, if we encounter an event that is less than .5 seconds want to
+    // pad the audio event until it is .5 seconds long
+    const minimumEventDuration = 0.5 as const;
+    const eventDuration = safeEventEndTime - roundedStartTime;
+    if (eventDuration < minimumEventDuration) {
+      const requiredPaddingAmount = minimumEventDuration - eventDuration;
+      const safeEventTimes = this.padAudioEvent(
+        safeEventStartTime,
+        safeEventEndTime,
+        requiredPaddingAmount,
+        audioRecording,
+      );
+
+      safeEventStartTime = safeEventTimes.startTimeSeconds;
+      safeEventEndTime = safeEventTimes.endTimeSeconds;
+    }
+
+    const basePath =
+      `/audio_recordings/${audioEvent.audioRecordingId}/media.flac`;
+    let urlParams = `?audio_event_id=${audioEvent.id}` +
+      `&start_offset=${safeEventStartTime}&end_offset=${safeEventEndTime}`;
```

Comment:
you've got most of the parts in place here but I'd suggest a few tweaks:

- accept padding as an argument (default 0)
- calculate midpoint first
- if duration < 1 && padding  == 0 
    - then set padding to (1 - duration)/2
- then pad
- then floor/ceil start/end
- then bound check
   - on over/under flow, take excess and add it to the other side